### PR TITLE
Restructure history service mutable state & related logics

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Rename mutablestate package to workflow package
* Rename history cache to workflow.Cache
* Rename workflowExecutionContext to workflow.Context
* Rename stateBuilder to workflow.MutableStateRebuilder
* Rename mutableStateTaskGenerator to workflow.TaskGenerato
* Rename mutableStateTaskRefresher to workflow.TaskRefresher
* Move mutableState to workflow package
* Move retry util to workflow package
* Move timer sequence util to workflow package
* Move query util to workflow package
* New tests package containing test related variables
* No business logic change; pure code restructure

**Why?**
Code restructure, pre-requisite to other code restructures

**How did you test it?**
Existing tests & CICD

**Potential risks**
N/A

**Is hotfix candidate?**
No

### CODE REVIEW

**Review by yiminc — APPROVED** [review]
diff tooooo looooooong...

File: service/history/historyEngine.go [hunk 1]
```diff
@@ -383,21 +336,21 @@ func (e *historyEngineImpl) registerNamespaceFailoverCallback() {
...
+	// 4. failover start & task processing Unlock & shard namespace version notification update. (order does not matter for this discussion)
 	//
 	// The above guarantees that task created during the failover will be processed.
 	// If the task is created after namespace change:
 	// 		then active processor will handle it. (simple case)
 	// If the task is created before namespace change:
-	//		task -> release of shard lock
-	//		failover min / max task levels calculated & updated to shard (using shard lock) -> failover start
+	//		task -> release of shard Lock
+	//		failover min / max task levels calculated & updated to shard (using shard Lock) -> failover start
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/historyEngine.go:353]: could be, use a lot of batch action provided by goland

**Comment by wxing1292** [comment]
> diff tooooo looooooong...

sorry, but due to potential circular dependency issue, i have to make all changes in one PR

File: service/history/consts/const.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	// ErrEmptyHistoryRawEventBatch indicate that one single batch of history raw events is of size 0
+	ErrEmptyHistoryRawEventBatch = serviceerror.NewInvalidArgument("encounter empty history batch")
+	// ErrSizeExceedsLimit is error indicating workflow execution has exceeded system defined limit
+	ErrSizeExceedsLimit = serviceerror.NewResourceExhausted(common.FailureReasonSizeExceedsLimit)
+	// ErrUnknownCluster is error indicating unknown cluster
+	ErrUnknownCluster = serviceerror.NewInvalidArgument("unknown cluster")
+
+	// FailedWorkflowStatuses is a set of failed workflow close states, used for start workflow policy
+	// for start workflow execution API
+	FailedWorkflowStatuses = map[enumspb.WorkflowExecutionStatus]bool{
```
Comment (alexshtin) [inline service/history/consts/const.go:78]:
```suggestion
	FailedWorkflowStatuses = map[enumspb.WorkflowExecutionStatus]struct{}{
```
?

File: service/history/consts/const.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
+// The MIT License
```
Comment (alexshtin) [inline service/history/consts/const.go:1]: `const` is the same as `util`, `common`, and so one. I am against packages with such names. I usually put all errors to `errors.go` and all other consts to `defs.go` but leave them within the same package (`history` in this case).

If it causes cycles, then it is good sign that more refactoring is needed.

File: service/history/nDCStandbyTaskUtil.go [hunk 4]
```diff
@@ -157,7 +159,7 @@ func newPushWorkflowTaskToMatchingInfo(
 }
 
 func getHistoryResendInfo(
-	mutableState mutableState,
+	mutableState workflow.MutableState,
```
Comment (alexshtin) [inline service/history/nDCStandbyTaskUtil.go:162]: This looks nice.

File: service/history/tests/const.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/cache"
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/service/history/configs"
+)
+
+var Version = int64(1234)
```
Comment (alexshtin) [inline service/history/tests/const.go:40]: single `var`?

File: service/history/tests/const.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
+// The MIT License
```
Comment (alexshtin) [inline service/history/tests/const.go:1]: We have `defs.go` in many places already. What to do you think about naming this file in the same way?

File: service/history/timerQueueStandbyTaskExecutor.go [hunk 7]
```diff
@@ -208,7 +210,7 @@ func (t *timerQueueStandbyTaskExecutor) executeActivityTimeoutTask(
 		// see comments at the beginning of this function.
 		// NOTE: this is the only place in the standby logic where mutable state can be updated
 
-		// need to clear the activity heartbeat timer task marks
+		// need to Clear the activity heartbeat timer task marks
```
Comment (alexshtin) [inline service/history/timerQueueStandbyTaskExecutor.go:213]: This is GoLand, I bet.

File: service/history/transferQueueActiveTaskExecutor.go [hunk 8]
```diff
@@ -667,8 +669,8 @@ func (t *transferQueueActiveTaskExecutor) processStartChildExecution(
 		return err
 	}
 
-	// NOTE: do not access anything related mutable state after this lock release
-	// release the context lock since we no longer need mutable state builder and
+	// NOTE: do not access anything related mutable state after this Lock release
+	// release the context Lock since we no longer need mutable state builder and
```
Comment (alexshtin) [inline service/history/transferQueueActiveTaskExecutor.go:673]: Same thing. There is a way in GoLand to search in comments only. I think you can easily revert these renames by searching for "Lock" and "Clear" case sensitive in comments only.

File: service/history/workflow/mutable_state_impl.go [hunk 9]
```diff
@@ -99,7 +99,7 @@ var (
 )
 
 type (
-	mutableStateBuilder struct {
+	MutableStateImpl struct {
```
Comment (alexshtin) [inline service/history/workflow/mutable_state_impl.go:102]: This name is definitely better than old one but I don't like `Impl` suffix. From the other side, I can't figure out the better name.

File: service/history/workflow/mutable_state_rebuilder.go [hunk 10]
```diff
@@ -41,30 +41,29 @@ import (
...
 	"go.temporal.io/server/service/history/shard"
 )
 
 type (
-	taskGeneratorProvider func(mutableState) mutableStateTaskGenerator
+	taskGeneratorProvider func(MutableState) TaskGenerator
 
-	stateBuilder interface {
-		applyEvents(
+	MutableStateRebuilder interface {
```
Comment (alexshtin) [inline service/history/workflow/mutable_state_rebuilder.go:50]: :+1:

File: service/history/workflow/workflow_task_state_machine.go [hunk 11]
```diff
@@ -159,34 +100,34 @@ func (m *mutableStateWorkflowTaskManagerImpl) ReplicateTransientWorkflowTaskSche
 	// this is OK
 	// 1. if a failover happen just after this transient workflow task,
 	// AddWorkflowTaskStartedEvent will handle the correction of schedule ID
-	// and set the attempt to 1
+	// and set the Attempt to 1
```
Comment (alexshtin) [inline service/history/workflow/workflow_task_state_machine.go:103]: and `Attempt`...

**Review by alexshtin — APPROVED** [review]

File: service/history/tests/const.go (hunk 6 above)
Comment (wxing1292) [reply service/history/tests/const.go:1]: this is not really definitions, but vars for tests. i will rename it to vars.go for this PR

File: service/history/consts/const.go (hunk 3 above)
Comment (wxing1292) [reply service/history/consts/const.go:1]: 1.  leaving the var definitions in history package will cause circular dependency. e.g. `history` package and `history/workflow` package will reference the same `history` package for errors, `history` package will reference `history/workflow` for internal logics
2. moving errors definitions into errors package will cause import conflict `"errors"` vs `"history/errors"`

let me know if your have other suggestions of names

Comment (alexshtin) [reply service/history/consts/const.go:1]: 1. This is what I am saying. More refactoring and decomposition is needed here. I can't tell what exactly but this should be a ultimate goal here.
2. I am not suggesting to add `history/errors` package. I suggest to move errors definitions to `errors.go` within `history` package. And again, it probably will cause cycles, but see (1).

File: service/history/tests/const.go (hunk 6 above)
Comment (alexshtin) [reply service/history/tests/const.go]: ie "definition of var for tests". It depends on how you think about it. I don't have strong opinion on file names. But I have strong opinion on consistency of those names. Let's make a call, figure out the name for such code (almost every package has it) and use it everywhere.

File: service/history/consts/const.go (hunk 3 above)
Comment (wxing1292) [reply service/history/consts/const.go:1]: 1. the goal is to make the code structure easy to read / modify / understand, not to making the code shine.
2. like i said, there will be circular dependency issue, thanks to golang

File: service/history/tests/const.go (hunk 6 above)
Comment (wxing1292) [reply service/history/tests/const.go]: > But I have strong opinion on consistency of those names.

i believe we should focus on things that matters? 
file names does not participate in imports (dir name do, but that is a different thing)
as long as file name itself gives an indication of what it contains, that should do the work

**Pull request merged**